server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -387,6 +389,7 @@ public static APIBlock readFrom(StreamInput input) throws IOException {
     private final ActiveShardCount waitForActiveShards;
     private final ImmutableOpenMap<String, RolloverInfo> rolloverInfos;
     private final boolean isSystem;
+    private final boolean isDataStreamIndex;
```

Comment:
I don't think we should change the index metadata for this information. That would only work on data streams created after 7.15 unless I missed something ?
Checking if a concrete index is a data stream can be done dynamically when the `IndexShard` or `IndexService` is created ? You can access the index abstraction lookup with `clusterService.state().metadata().getIndicesLookup()`.

Reply:
+1 to what @jimczi said here. Individual indices can be detached from and added to data streams in a variety of scenarios including snapshot restores the relatively new [Migrate to Data Stream API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-migrate-to-data-stream.html). If such a flag were included in index metadata, code would need to be added in all such places to correct the index metadata and we'd prefer to avoid that.

Reply:
@jimczi  Thank you for comment. I've started with a dynamic lookup `clusterService.state().metadata().getIndicesLookup()` like you suggested but it is not allowed to to access a cluster state while modifying cluster state (e.g. during index creation). There is a function `ClusterApplierService::assertNotCalledFromClusterStateApplier` that checks for it, and returns an error: "the applied cluster state is not yet available". I can investigate if it is possible to do this check somewhere earlier.

@danhermann Thanks for checking, and highlighting Migrate Data Stream API. I would think that this API would be quite rare to use, and most datastream indices would be created in a usual way.  But indeed if used, we would need to update  index metadata.

I will see if it is possible to do this check dynamically,  but as we build more code around datastreams it would be nice to have an easy way to check if a shard is a part of datastream, with APIs like:   `IndexMetadata::isDataStreamIndex` or `IndexShard::isDataStreamIndex`.